src/Microsoft.Extensions.ServiceDiscovery/ServiceEndpointWatcher.cs

```diff
@@ -244,10 +253,17 @@ private void SchedulePollingTimer()
     /// <inheritdoc/>
     public async ValueTask DisposeAsync()
     {
-        lock (_lock)
+        try
         {
             _disposalCancellation.Cancel();
+        }
+        catch (Exception exception)
+        {
+            _logger.LogError(exception, "Error cancelling disposal cancellation token.");
```

Comment:
Just curious, what can cause this error

Reply:
@danmoseley When you call `CancellationTokenSource.Cancel()`, exceptions thrown by the callbacks are propagated to the caller. See this doc comment fragment:
```xml
/// <exception cref="AggregateException">An aggregate exception containing all the exceptions thrown
/// by the registered callbacks on the associated <see cref="CancellationToken"/>.</exception>
```

The online docs also have this remark:
> We recommend that cancelable operations and callbacks registered with [CancellationToken](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=net-8.0) not throw exceptions.

But, of course, we cannot guarantee that.

Reply:
The log line could say "error thrown by cancellation callback", but that would be presumptuous

Reply:
Is it internal? The token?

Reply:
We pass the cancellation token to providers, and we don't know what they will do since that is an extensibility point.